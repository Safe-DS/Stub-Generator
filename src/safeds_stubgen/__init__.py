"""Safe-DS stubs generator."""
from __future__ import annotations
