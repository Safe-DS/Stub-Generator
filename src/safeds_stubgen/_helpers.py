def is_internal(name: str) -> bool:
    return name.startswith("_")
