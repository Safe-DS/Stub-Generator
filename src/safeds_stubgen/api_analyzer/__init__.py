"""API-Analyzer for the Safe-DS stubs generator."""
from __future__ import annotations

from ._api import API, Attribute, Class, Function, Parameter, ParameterAssignment
from ._get_api import get_api
from ._mypy_helpers import get_classdef_definitions, get_funcdef_definitions, get_mypyfile_definitions
from ._package_metadata import distribution, distribution_version, package_root
from ._types import (
    AbstractType,
    BoundaryType,
    DictType,
    EnumType,
    FinalType,
    ListType,
    LiteralType,
    NamedType,
    OptionalType,
    SetType,
    TupleType,
    UnionType,
)

__all__ = [
    "AbstractType",
    "API",
    "Attribute",
    "BoundaryType",
    "Class",
    "DictType",
    "distribution",
    "distribution_version",
    "EnumType",
    "FinalType",
    "Function",
    "get_api",
    "get_classdef_definitions",
    "get_funcdef_definitions",
    "get_mypyfile_definitions",
    "ListType",
    "LiteralType",
    "NamedType",
    "OptionalType",
    "package_root",
    "Parameter",
    "ParameterAssignment",
    "SetType",
    "TupleType",
    "UnionType",
]
