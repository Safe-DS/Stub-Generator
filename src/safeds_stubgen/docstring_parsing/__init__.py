"""Parsing docstrings into a common format."""

from ._abstract_docstring_parser import AbstractDocstringParser
from ._create_docstring_parser import create_docstring_parser
from ._docstring import AttributeDocstring, ClassDocstring, FunctionDocstring, ParameterDocstring, ResultDocstring
from ._docstring_style import DocstringStyle
from ._epydoc_parser import EpydocParser
from ._googledoc_parser import GoogleDocParser
from ._numpydoc_parser import NumpyDocParser
from ._plaintext_docstring_parser import PlaintextDocstringParser
from ._restdoc_parser import RestDocParser

__all__ = [
    "AbstractDocstringParser",
    "AttributeDocstring",
    "ClassDocstring",
    "create_docstring_parser",
    "DocstringStyle",
    "EpydocParser",
    "FunctionDocstring",
    "GoogleDocParser",
    "NumpyDocParser",
    "ParameterDocstring",
    "PlaintextDocstringParser",
    "RestDocParser",
    "ResultDocstring",
]
