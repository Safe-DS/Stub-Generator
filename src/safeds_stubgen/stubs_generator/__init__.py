"""API-Analyzer for the Safe-DS stubs generator."""

from __future__ import annotations

from ._generate_stubs import create_stub_files, generate_stub_data
from ._helper import NamingConvention
from ._stub_string_generator import StubsStringGenerator

__all__ = [
    "create_stub_files",
    "generate_stub_data",
    "NamingConvention",
    "StubsStringGenerator",
]
