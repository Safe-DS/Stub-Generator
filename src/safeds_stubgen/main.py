"""The entrypoint to the program."""
from __future__ import annotations

import time

from safeds_stubgen.api_analyzer.cli import cli


def main() -> None:
    """Launch the program."""
    start_time = time.time()

    cli()

    print("\n============================================================")  # noqa: T201
    print(f"Program ran in {time.time() - start_time}s")  # noqa: T201


if __name__ == "__main__":
    main()
