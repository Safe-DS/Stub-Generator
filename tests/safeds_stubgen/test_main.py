import json
import sys
from pathlib import Path

import pytest
from safeds_stubgen.main import main
from syrupy.assertion import SnapshotAssertion

_lib_dir = Path(__file__).parent.parent.parent
_test_package_name = "main_package"
_test_package_name_boundaries_valid_values_numpydoc = "boundary_enum_package_numpydoc"
_test_package_name_boundaries_valid_values_googledoc = "boundary_enum_package_googledoc"
_test_package_name_boundaries_valid_values_restdoc = "boundary_enum_package_restdoc"
_test_package_purity = "purity_package"
_main_dir = Path(_lib_dir / "src" / "main.py")
_test_package_dir = Path(_lib_dir / "tests" / "data" / _test_package_name)
_out_dir = Path(_lib_dir / "tests" / "data" / "out")

_astroid_evaluation_dir = Path(_lib_dir / "tests" / "data" / "astroid")
_safeds_v19_evaluation_dir = Path(_lib_dir / "tests" / "data" / "safeds")

@pytest.mark.parametrize(
    ("test_package_name", "out_file_dir", "docstyle"),
    [
        (
            _test_package_name,
            Path(_out_dir / f"{_test_package_name}__api.json"),
            "plaintext"
        ),
        (
            _test_package_name_boundaries_valid_values_numpydoc,
            Path(_out_dir / f"{_test_package_name_boundaries_valid_values_numpydoc}__api.json"),
            "numpydoc"
        ),
        (
            _test_package_name_boundaries_valid_values_googledoc,
            Path(_out_dir / f"{_test_package_name_boundaries_valid_values_googledoc}__api.json"),
            "google"
        ),
        (
            _test_package_name_boundaries_valid_values_restdoc,
            Path(_out_dir / f"{_test_package_name_boundaries_valid_values_restdoc}__api.json"),
            "rest"
        ),
        (
            _test_package_purity,
            Path(_out_dir / f"{_test_package_purity}__api_purity.json"),
            "numpydoc"
        ),
    ],
    ids=[
        "plaintext",
        "numpydoc - boundary - enum",
        "googledoc - boundary - enum",
        "restdoc - boundary - enum",
        "purity"
    ],
)
def test_main(
    test_package_name: str,
    out_file_dir: Path,
    docstyle: str,
    snapshot: SnapshotAssertion
) -> None:
    # Overwrite system arguments

    sys.argv = [
        str(_main_dir),
        "-v",
        "-s",
        str(Path(_lib_dir / "tests" / "data" / test_package_name)),
        "-o",
        str(_out_dir),
        "-tr",
        "--docstyle",
        docstyle,
        "-nc",
    ]

    main()

    with Path.open(out_file_dir, encoding="utf-8") as f:
        json_data = json.load(f)

    assert json_data == snapshot

@pytest.mark.parametrize(
    ("package_path", "out_file_dir", "docstyle"),
    [
        (
            _astroid_evaluation_dir,
            Path(_out_dir / f"{'astroid'}__api_purity.json"),
            "numpydoc"
        ),
        (
            _safeds_v19_evaluation_dir,
            Path(_out_dir / f"{'safedsV19'}__api_purity.json"),
            "numpydoc"
        ),

    ],
    ids=[
        "astroid",
        "safedsV19"
    ],
)
def test_evaluation(
    package_path: Path,
    out_file_dir: Path,
    docstyle: str,
    snapshot: SnapshotAssertion
) -> None:
    # Overwrite system arguments

    sys.argv = [
        str(_main_dir),
        "-v",
        "-s",
        str(package_path),
        "-o",
        str(_out_dir),
        "-tr",
        "--docstyle",
        docstyle,
        "-nc",
        "--evaluate_purity",
        # "--evaluate_api",
        "-old"
    ]

    main()
    assert True

def test_main_empty() -> None:
    # Overwrite system arguments
    sys.argv = [
        str(_main_dir),
        "-v",
        "-s",
        str(_test_package_dir),
        "-o",
        str(_out_dir),
        "--docstyle",
        "plaintext",
    ]

    with pytest.raises(ValueError, match="No files found to analyse."):
        main()
