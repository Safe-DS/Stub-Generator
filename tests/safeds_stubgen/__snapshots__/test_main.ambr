# serializer version: 1
# name: test_main
  dict({
    'attributes': list([
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_init_attr_private',
        'is_public': False,
        'is_static': False,
        'is_type_inferred': False,
        'name': '_init_attr_private',
        'type': dict({
          'kind': 'NamedType',
          'name': 'float',
          'qname': '',
        }),
      }),
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/attr_1',
        'is_public': True,
        'is_static': True,
        'is_type_inferred': False,
        'name': 'attr_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'int',
          'qname': '',
        }),
      }),
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/init_attr',
        'is_public': True,
        'is_static': False,
        'is_type_inferred': False,
        'name': 'init_attr',
        'type': dict({
          'kind': 'NamedType',
          'name': 'bool',
          'qname': '',
        }),
      }),
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/NestedPrivateClass/nested_class_attr',
        'is_public': False,
        'is_static': True,
        'is_type_inferred': True,
        'name': 'nested_class_attr',
        'type': dict({
          'kind': 'NamedType',
          'name': 'int',
          'qname': '',
        }),
      }),
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/public_attr_in_private_class',
        'is_public': False,
        'is_static': True,
        'is_type_inferred': True,
        'name': 'public_attr_in_private_class',
        'type': dict({
          'kind': 'NamedType',
          'name': 'int',
          'qname': '',
        }),
      }),
      dict({
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/public_init_attr_in_private_class',
        'is_public': False,
        'is_static': False,
        'is_type_inferred': False,
        'name': 'public_init_attr_in_private_class',
        'type': dict({
          'kind': 'NamedType',
          'name': 'int',
          'qname': '',
        }),
      }),
    ]),
    'classes': list([
      dict({
        'attributes': list([
        ]),
        'classes': list([
        ]),
        'constructor': None,
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/another_path/another_module/AnotherClass',
        'is_public': True,
        'methods': list([
        ]),
        'name': 'AnotherClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
        ]),
        'variances': list([
        ]),
      }),
      dict({
        'attributes': list([
          'test_main/main_test_module/ModuleClass/attr_1',
          'test_main/main_test_module/ModuleClass/init_attr',
          'test_main/main_test_module/ModuleClass/_init_attr_private',
        ]),
        'classes': list([
          'test_main/main_test_module/ModuleClass/NestedClass',
        ]),
        'constructor': dict({
          'docstring': dict({
            'description': '''
              Summary of the init description.
              
              Full init description.
            ''',
            'full_docstring': '''
              Summary of the init description.
              
              Full init description.
            ''',
          }),
          'id': 'test_main/main_test_module/ModuleClass/__init__',
          'is_class_method': False,
          'is_property': False,
          'is_public': True,
          'is_static': False,
          'name': '__init__',
          'parameters': list([
            'test_main/main_test_module/ModuleClass/__init__/self',
            'test_main/main_test_module/ModuleClass/__init__/init_param_1',
          ]),
          'reexported_by': list([
          ]),
          'results': list([
          ]),
        }),
        'docstring': dict({
          'description': '''
            Summary of the description.
            
            Full description
          ''',
          'full_docstring': '''
            Summary of the description.
            
            Full description
          ''',
        }),
        'id': 'test_main/main_test_module/ModuleClass',
        'is_public': True,
        'methods': list([
          'test_main/main_test_module/ModuleClass/_some_function',
        ]),
        'name': 'ModuleClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
          'tests.data.test_main.main_test_module.AcDoubleAlias',
        ]),
        'variances': list([
        ]),
      }),
      dict({
        'attributes': list([
        ]),
        'classes': list([
        ]),
        'constructor': None,
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/NestedClass',
        'is_public': True,
        'methods': list([
          'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function',
        ]),
        'name': 'NestedClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
          'tests.data.test_main.another_path.another_module.AnotherClass',
        ]),
        'variances': list([
        ]),
      }),
      dict({
        'attributes': list([
          'test_main/main_test_module/_PrivateClass/public_attr_in_private_class',
          'test_main/main_test_module/_PrivateClass/public_init_attr_in_private_class',
        ]),
        'classes': list([
          'test_main/main_test_module/_PrivateClass/NestedPrivateClass',
        ]),
        'constructor': dict({
          'docstring': dict({
            'description': '',
            'full_docstring': '',
          }),
          'id': 'test_main/main_test_module/_PrivateClass/__init__',
          'is_class_method': False,
          'is_property': False,
          'is_public': False,
          'is_static': False,
          'name': '__init__',
          'parameters': list([
            'test_main/main_test_module/_PrivateClass/__init__/self',
          ]),
          'reexported_by': list([
          ]),
          'results': list([
          ]),
        }),
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass',
        'is_public': False,
        'methods': list([
          'test_main/main_test_module/_PrivateClass/public_func_in_private_class',
        ]),
        'name': '_PrivateClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
        ]),
        'variances': list([
        ]),
      }),
      dict({
        'attributes': list([
          'test_main/main_test_module/_PrivateClass/NestedPrivateClass/nested_class_attr',
        ]),
        'classes': list([
          'test_main/main_test_module/_PrivateClass/NestedPrivateClass/NestedNestedPrivateClass',
        ]),
        'constructor': None,
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/NestedPrivateClass',
        'is_public': True,
        'methods': list([
          'test_main/main_test_module/_PrivateClass/NestedPrivateClass/static_nested_private_class_function',
        ]),
        'name': 'NestedPrivateClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
        ]),
        'variances': list([
        ]),
      }),
      dict({
        'attributes': list([
        ]),
        'classes': list([
        ]),
        'constructor': None,
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/NestedPrivateClass/NestedNestedPrivateClass',
        'is_public': True,
        'methods': list([
        ]),
        'name': 'NestedNestedPrivateClass',
        'reexported_by': list([
        ]),
        'superclasses': list([
        ]),
        'variances': list([
        ]),
      }),
    ]),
    'distribution': '',
    'enum_instances': list([
    ]),
    'enums': list([
    ]),
    'functions': list([
      dict({
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function',
        'is_class_method': False,
        'is_property': False,
        'is_public': True,
        'is_static': False,
        'name': 'nested_class_function',
        'parameters': list([
          'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/self',
          'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/param_1',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
          'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/result_1',
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '''
            Summary of the init description.
            
            Full init description.
          ''',
          'full_docstring': '''
            Summary of the init description.
            
            Full init description.
          ''',
        }),
        'id': 'test_main/main_test_module/ModuleClass/__init__',
        'is_class_method': False,
        'is_property': False,
        'is_public': True,
        'is_static': False,
        'name': '__init__',
        'parameters': list([
          'test_main/main_test_module/ModuleClass/__init__/self',
          'test_main/main_test_module/ModuleClass/__init__/init_param_1',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '''
            Function Docstring.
            
            param_2: bool.
          ''',
          'full_docstring': '''
            Function Docstring.
            
            param_2: bool.
          ''',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_some_function',
        'is_class_method': False,
        'is_property': False,
        'is_public': False,
        'is_static': False,
        'name': '_some_function',
        'parameters': list([
          'test_main/main_test_module/ModuleClass/_some_function/self',
          'test_main/main_test_module/ModuleClass/_some_function/param_1',
          'test_main/main_test_module/ModuleClass/_some_function/param_2',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
          'test_main/main_test_module/ModuleClass/_some_function/result_1',
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/NestedPrivateClass/static_nested_private_class_function',
        'is_class_method': False,
        'is_property': False,
        'is_public': False,
        'is_static': True,
        'name': 'static_nested_private_class_function',
        'parameters': list([
        ]),
        'reexported_by': list([
        ]),
        'results': list([
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/__init__',
        'is_class_method': False,
        'is_property': False,
        'is_public': False,
        'is_static': False,
        'name': '__init__',
        'parameters': list([
          'test_main/main_test_module/_PrivateClass/__init__/self',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/public_func_in_private_class',
        'is_class_method': False,
        'is_property': False,
        'is_public': False,
        'is_static': False,
        'name': 'public_func_in_private_class',
        'parameters': list([
          'test_main/main_test_module/_PrivateClass/public_func_in_private_class/self',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'full_docstring': '',
        }),
        'id': 'test_main/main_test_module/_private_global_func',
        'is_class_method': False,
        'is_property': False,
        'is_public': False,
        'is_static': False,
        'name': '_private_global_func',
        'parameters': list([
        ]),
        'reexported_by': list([
        ]),
        'results': list([
          'test_main/main_test_module/_private_global_func/result_1',
        ]),
      }),
      dict({
        'docstring': dict({
          'description': '''
            Docstring 1.
            
            Docstring 2.
          ''',
          'full_docstring': '''
            Docstring 1.
            
            Docstring 2.
          ''',
        }),
        'id': 'test_main/main_test_module/global_func',
        'is_class_method': False,
        'is_property': False,
        'is_public': True,
        'is_static': False,
        'name': 'global_func',
        'parameters': list([
          'test_main/main_test_module/global_func/param_1',
          'test_main/main_test_module/global_func/param_2',
        ]),
        'reexported_by': list([
        ]),
        'results': list([
          'test_main/main_test_module/global_func/result_1',
        ]),
      }),
    ]),
    'modules': list([
      dict({
        'classes': list([
          'test_main/another_path/another_module/AnotherClass',
        ]),
        'docstring': '''
          Another Module Docstring.
          
          Full Docstring Description
  
        ''',
        'enums': list([
        ]),
        'functions': list([
        ]),
        'id': 'test_main/another_path/another_module',
        'name': 'another_module',
        'qualified_imports': list([
        ]),
        'wildcard_imports': list([
        ]),
      }),
      dict({
        'classes': list([
          'test_main/main_test_module/ModuleClass',
          'test_main/main_test_module/_PrivateClass',
        ]),
        'docstring': 'Docstring of the some_class.py module.',
        'enums': list([
        ]),
        'functions': list([
          'test_main/main_test_module/global_func',
          'test_main/main_test_module/_private_global_func',
        ]),
        'id': 'test_main/main_test_module',
        'name': 'main_test_module',
        'qualified_imports': list([
          dict({
            'alias': 'mathematics',
            'qualified_name': 'math',
          }),
          dict({
            'alias': None,
            'qualified_name': 'mypy',
          }),
          dict({
            'alias': None,
            'qualified_name': 'another_path.another_module.AnotherClass',
          }),
          dict({
            'alias': '_AcImportAlias',
            'qualified_name': 'another_path.another_module.AnotherClass',
          }),
        ]),
        'wildcard_imports': list([
          dict({
            'module_name': 'docstring_parser',
          }),
        ]),
      }),
    ]),
    'package': 'test_main',
    'parameters': list([
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/param_1',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'param_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'int',
          'qname': '',
        }),
      }),
      dict({
        'assigned_by': 'IMPLICIT',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/self',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'self',
        'type': None,
      }),
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/__init__/init_param_1',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'init_param_1',
        'type': None,
      }),
      dict({
        'assigned_by': 'IMPLICIT',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/__init__/self',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'self',
        'type': None,
      }),
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_some_function/param_1',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'param_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'AnotherClass',
          'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
        }),
      }),
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': False,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_some_function/param_2',
        'is_optional': True,
        'is_type_inferred': False,
        'name': 'param_2',
        'type': dict({
          'kind': 'NamedType',
          'name': 'bool',
          'qname': '',
        }),
      }),
      dict({
        'assigned_by': 'IMPLICIT',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_some_function/self',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'self',
        'type': None,
      }),
      dict({
        'assigned_by': 'IMPLICIT',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/__init__/self',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'self',
        'type': None,
      }),
      dict({
        'assigned_by': 'IMPLICIT',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_PrivateClass/public_func_in_private_class/self',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'self',
        'type': None,
      }),
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': 'first param',
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/global_func/param_1',
        'is_optional': True,
        'is_type_inferred': False,
        'name': 'param_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'str',
          'qname': '',
        }),
      }),
      dict({
        'assigned_by': 'POSITION_OR_NAME',
        'default_value': None,
        'docstring': dict({
          'default_value': '',
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/global_func/param_2',
        'is_optional': False,
        'is_type_inferred': False,
        'name': 'param_2',
        'type': dict({
          'kind': 'UnionType',
          'types': list([
            dict({
              'kind': 'NamedType',
              'name': 'AnotherClass',
              'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
            }),
            dict({
              'kind': 'NamedType',
              'name': 'None',
              'qname': '',
            }),
          ]),
        }),
      }),
    ]),
    'results': list([
      dict({
        'docstring': dict({
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/NestedClass/nested_class_function/result_1',
        'is_type_inferred': False,
        'name': 'result_1',
        'type': dict({
          'kind': 'SetType',
          'types': list([
            dict({
              'kind': 'UnionType',
              'types': list([
                dict({
                  'kind': 'NamedType',
                  'name': 'bool',
                  'qname': '',
                }),
                dict({
                  'kind': 'NamedType',
                  'name': 'None',
                  'qname': '',
                }),
              ]),
            }),
          ]),
        }),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/ModuleClass/_some_function/result_1',
        'is_type_inferred': False,
        'name': 'result_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'AnotherClass',
          'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
        }),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/_private_global_func/result_1',
        'is_type_inferred': False,
        'name': 'result_1',
        'type': dict({
          'kind': 'UnionType',
          'types': list([
            dict({
              'kind': 'NamedType',
              'name': 'AnotherClass',
              'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
            }),
            dict({
              'kind': 'NamedType',
              'name': 'AnotherClass',
              'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
            }),
            dict({
              'kind': 'NamedType',
              'name': 'AnotherClass',
              'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
            }),
          ]),
        }),
      }),
      dict({
        'docstring': dict({
          'description': '',
          'type': '',
        }),
        'id': 'test_main/main_test_module/global_func/result_1',
        'is_type_inferred': False,
        'name': 'result_1',
        'type': dict({
          'kind': 'NamedType',
          'name': 'AnotherClass',
          'qname': 'tests.data.test_main.another_path.another_module.AnotherClass',
        }),
      }),
    ]),
    'schemaVersion': 1,
    'version': '',
  })
# ---
